### PR TITLE
[SPARK-39176][PYSPARK] Fixed a problem with pyspark serializing pre-1970 datetime in windows

### PR DESCRIPTION
**Pull request by dingsl-giser**
### What changes were proposed in this pull request?

Fix problems with pyspark in Windows:
1. Fixed datetime conversion to timestamp before 1970;
2. Fixed datetime conversion when timestamp is negative;
3. Adding a test script.

### Why are the changes needed?

Pyspark has problems serializing pre-1970 times in Windows.

An exception occurs when executing the following code under Windows:
```python
rdd = sc.parallelize([('a', datetime(1957, 1, 9, 0, 0)),
                      ('b', datetime(2014, 1, 27, 0, 0))])
df = spark.createDataFrame(rdd, ["id", "date"])

df.show()
df.printSchema()

print(df.collect())
```

```bash
  File "...\spark\python\lib\pyspark.zip\pyspark\sql\types.py", line 195, in toInternal
    else time.mktime(dt.timetuple()))
OverflowError: mktime argument out of range

at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:503)
	at org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:638)
	at org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:621)
	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:456)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:489)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:729)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:340)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:872)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:872)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:349)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:313)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:127)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
```
and
```bash
File ...\spark\python\lib\pyspark.zip\pyspark\sql\types.py, in fromInternal:
Line 207:   return datetime.datetime.fromtimestamp(ts // 1000000).replace(microsecond=ts % 1000000)

OSError: [Errno 22] Invalid argument
```
 After updating the code, the above code was run successfully！
```bash
+---+-------------------+
| id|               date|
+---+-------------------+
|  a|1957-01-08 16:00:00|
|  b|2014-01-26 16:00:00|
+---+-------------------+

root
 |-- id: string (nullable = true)
 |-- date: timestamp (nullable = true)

[Row(id='a', date=datetime.datetime(1957, 1, 9, 0, 0)), Row(id='b', date=datetime.datetime(2014, 1, 27, 0, 0))] 
```
### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New and existing test suites

### CODE REVIEW

**Comment by AnywalkerGiser** [comment]
@HyukjinKwon I closed the RP in the 3.0 branch(https://github.com/apache/spark/pull/36537) and raised a new RP in the master branch.

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -212,15 +213,29 @@ def needConversion(self) -> bool:
 
     def toInternal(self, dt: datetime.datetime) -> int:
         if dt is not None:
-            seconds = (
-                calendar.timegm(dt.utctimetuple()) if dt.tzinfo else time.mktime(dt.timetuple())
-            )
+            if platform.system().lower() == 'windows':
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:216]: Should we also check if the value is negative?

**Comment by HyukjinKwon** [comment]
@AnywalkerGiser BTW, it would be easier to follow and review if you have any link on the OS difference in that Python library.

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (AnywalkerGiser) [reply python/pyspark/sql/types.py:216]: Don't need, this method is still available when the negative.

**Comment by AnywalkerGiser** [comment]
@HyukjinKwon Do you mean the platform library, added to the code comments?

**Comment by HyukjinKwon** [comment]
I am surprised that the same Python library returns a different value depending on OS. I know some libraries are dependent on C library implementation but didn't expect such drastic difference. Code comment or documentation anything is fine. I would like to see if this is an official difference in Python.

**Comment by AnywalkerGiser** [comment]
@HyukjinKwon [Python3 datetime interface documentation](https://docs.python.org/3/library/datetime.html)
This is explained in this document.
image

**Comment by HyukjinKwon** [comment]
So is C localtime function has a different behaviour in OS? that returns negative values?

**Comment by AnywalkerGiser** [comment]
The localtime function does not differ in OS, the fromtimestamp function does, and datetime many functions have problems resolving dates before 1970 in windows.

**Comment by AnywalkerGiser** [comment]
@HyukjinKwon Can this solution be merged into master?

**Comment by HyukjinKwon** [comment]
I would like to understand the problem first, and see if this behaviour difference is internal or official. This is core code path so would have to expect more reviews and time.

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/types.py:216]: How about performance? This is hot path that's executed every value, and the performance here is pretty critical.

**Comment by AnywalkerGiser** [comment]
Understood, looking forward to your reply later.

**Comment by HyukjinKwon** [comment]
@AnywalkerGiser mind pointing out any documentation that states this negative value?

**Comment by AnywalkerGiser** [comment]
Here are some blogs on related issues:
[Python | mktime overflow error](https://stackoverflow.com/questions/2518706/python-mktime-overflow-error)
[Python fromtimestamp OSError](https://stackoverflow.com/questions/37494983/python-fromtimestamp-oserror)

**Comment by AnywalkerGiser** [comment]
I think the Spark project can look for a solution if Python doesn't fix this bug.

**Comment by HyukjinKwon** [comment]
Does that happen in all Windows with all Python versions?

**Comment by HyukjinKwon** [comment]
is this a bug in Python?

**Comment by AnywalkerGiser** [comment]
I thought it was a bug in python, but the documentation said it would report an error if it was out of time range. I tested python 3.6, 3.7, and 3.8 on windows.

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -212,15 +213,32 @@ def needConversion(self) -> bool:
 
     def toInternal(self, dt: datetime.datetime) -> int:
         if dt is not None:
-            seconds = (
-                calendar.timegm(dt.utctimetuple()) if dt.tzinfo else time.mktime(dt.timetuple())
-            )
+            if platform.system().lower() == "windows":
```
Comment (srowen) [inline python/pyspark/sql/types.py:216]: I'm just super surprised this is Windows specific?

Comment (AnywalkerGiser) [reply python/pyspark/sql/types.py:216]: Yes, and I want you to check it out for yourself.

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (AnywalkerGiser) [reply python/pyspark/sql/types.py]: Are there professional performance testers?

**Comment by dingsl-giser** [comment]
@HyukjinKwon This problem still exists in the new version. Can it be merged?

**Pull request closed without merging**